# Proyecto Integrador - Minería de Datos I

## Notebook 01 - Inspección Inicial del Dataset

### Objetivo

El objetivo de este notebook es realizar una inspección inicial del conjunto de datos con el fin de comprender su estructura, identificar los tipos de variables, detectar posibles problemas de calidad y obtener evidencia que permita justificar las decisiones de preparación de datos en las siguientes etapas del proyecto.

Durante esta fase no se realizarán modificaciones sobre el dataset original. Únicamente se llevará a cabo una exploración descriptiva que servirá como base para el proceso de limpieza y análisis posterior.

## Índice

1. Importación de librerías
2. Carga del dataset
3. Exploración inicial
4. Dimensiones del dataset
5. Identificación de las variables
6. Información general del DataFrame
7. Estadísticas descriptivas
8. Análisis de valores faltantes
9. Identificación de registros duplicados
10. Análisis de las variables categóricas
11. Resumen de problemas detectados
12. Conclusiones

## 1. Importación de librerías

En esta primera etapa se importan las librerías que serán utilizadas durante la inspección inicial del dataset.

Las librerías permiten acceder a funciones específicas para la manipulación de datos, operaciones numéricas y visualización de información. Importarlas al comienzo del notebook facilita la organización del código y evita repetir instrucciones en distintas secciones del análisis.

In [ ]:
# Manipulación y análisis de datos
import pandas as pd
import numpy as np

# Visualización de datos
import matplotlib.pyplot as plt
import seaborn as sns

# Configuración estética de los gráficos
sns.set_style("whitegrid")

# Configuración para mostrar todas las columnas del DataFrame
pd.set_option('display.max_columns', None)

## 2. Carga del dataset

Una vez importadas las librerías necesarias, el siguiente paso consiste en cargar el conjunto de datos que será utilizado durante el proyecto.

El dataset se encuentra almacenado en formato JSON, por lo que se utilizará la función `read_json()` de la librería Pandas para convertir su contenido en un DataFrame.

En esta etapa no se realizará ninguna modificación sobre los datos originales. El objetivo es únicamente cargar el archivo para comenzar su inspección.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 3. Carga del conjunto de datos

Una vez establecida la conexión con Google Drive, se procede a cargar el archivo JSON proporcionado por la cátedra.

El archivo será leído mediante la función `read_json()` de la librería Pandas y almacenado en un DataFrame denominado `df`. Este DataFrame será la estructura principal sobre la cual se realizará la inspección y el análisis del conjunto de datos.

Durante esta etapa no se realizarán modificaciones sobre los registros originales.

In [ ]:
# Ruta del archivo JSON
ruta = "/content/drive/MyDrive/PI_Mineria_Datos_1/data/raw/streaming_users_dirty.json"

# Cargar el dataset
df = pd.read_json(ruta)

# Mostrar mensaje de confirmación
print("✅ Dataset cargado correctamente.")

✅ Dataset cargado correctamente.


## 4. Exploración inicial del dataset

Antes de realizar cualquier análisis es conveniente observar una muestra de los datos.

La función `head()` permite visualizar los primeros registros del DataFrame, facilitando una primera comprensión de la estructura del conjunto de datos, las variables disponibles y el formato en el que se encuentra almacenada la información.

Esta inspección inicial permite detectar posibles inconsistencias de forma preliminar, aunque en esta etapa aún no se tomarán decisiones de limpieza.

In [ ]:
# Visualizar los primeros registros del dataset
df.head()

,user_id,age,subscription_plan,monthly_watch_time_mins,country,favorite_genre,last_login_date,customer_support_tickets
0,10000,39,Estándar,805.8,Brasil,Crime,2025-03-04,99
1,10001,37,Estándar,1173.4,Colombia,Crime,2019-04-02,2
2,10002,28,Básico,401.0,Colombia,Crime,2018-04-13,0
3,10003,43,Básico,62.4,Uruguay,Thriller,2021-01-31,0
4,10004,51,Básico,477.8,Perú,Thriller,2020-09-30,1


### Interpretación

La visualización de los primeros registros permitió obtener una primera impresión del conjunto de datos. Se observa que cada fila representa un usuario de la plataforma de streaming, mientras que las columnas contienen información demográfica, características de la suscripción, hábitos de consumo y variables relacionadas con la actividad del usuario.

Esta primera inspección servirá como punto de partida para analizar con mayor detalle la estructura y la calidad del dataset.

## 5. Dimensiones del dataset

En esta sección se determinará el tamaño del conjunto de datos mediante la propiedad `shape` de Pandas.

Conocer la cantidad de filas y columnas es un paso fundamental durante la inspección inicial, ya que permite comprender la magnitud del dataset y anticipar el alcance del análisis que se realizará en las etapas posteriores.

In [ ]:
# Obtener la cantidad de filas y columnas
filas, columnas = df.shape

print(f"Cantidad de filas: {filas}")
print(f"Cantidad de columnas: {columnas}")

Cantidad de filas: 8160
Cantidad de columnas: 8


### Interpretación

El conjunto de datos está compuesto por **8160 registros** y **8 variables**.

La cantidad de registros representa el número de usuarios incluidos en el dataset, mientras que las variables contienen la información disponible para cada uno de ellos.

Conocer estas dimensiones resulta fundamental para comprender el alcance del análisis y sirve como referencia para verificar posteriormente si las tareas de limpieza modifican la cantidad de registros o variables.

## 6. Identificación de las variables

En esta sección se visualizarán los nombres de todas las variables presentes en el conjunto de datos.

Conocer las variables disponibles permite comprender qué tipo de información contiene el dataset y facilita la planificación de los análisis que se realizarán en las siguientes etapas del proyecto.

In [ ]:
# Mostrar el nombre de todas las columnas
for columna in df.columns:
    print(columna)

user_id
age
subscription_plan
monthly_watch_time_mins
country
favorite_genre
last_login_date
customer_support_tickets


### Descripción de las variables

La siguiente tabla resume las variables presentes en el conjunto de datos y su significado dentro del contexto del proyecto.

| Variable | Tipo | Descripción |
|----------|------|-------------|
| **user_id** | Identificador | Identificador único de cada usuario. |
| **age** | Numérica | Edad del usuario. |
| **subscription_plan** | Categórica | Plan de suscripción contratado por el usuario. |
| **monthly_watch_time_mins** | Numérica | Cantidad de minutos visualizados durante el mes. |
| **country** | Categórica | País de residencia del usuario. |
| **favorite_genre** | Categórica | Género audiovisual preferido por el usuario. |
| **last_login_date** | Fecha | Fecha correspondiente al último acceso registrado del usuario. |
| **customer_support_tickets** | Numérica | Cantidad de consultas o tickets realizados al servicio de atención al cliente. |

## 7. Información general del DataFrame

En esta etapa se utilizará la función `info()` de Pandas para obtener un resumen general del conjunto de datos.

Esta función proporciona información relevante sobre cada variable, incluyendo su tipo de dato, la cantidad de valores no nulos y el uso aproximado de memoria.

La información obtenida permitirá identificar posibles problemas de calidad de datos que deberán analizarse en las siguientes etapas del proyecto.

In [ ]:
# Mostrar información general del DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8160 entries, 0 to 8159
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   user_id                   8160 non-null   int64  
 1   age                       8160 non-null   int64  
 2   subscription_plan         8160 non-null   object 
 3   monthly_watch_time_mins   7967 non-null   float64
 4   country                   8160 non-null   object 
 5   favorite_genre            7920 non-null   object 
 6   last_login_date           7840 non-null   object 
 7   customer_support_tickets  8160 non-null   int64  
dtypes: float64(1), int64(3), object(4)
memory usage: 510.1+ KB


### Interpretación

La función `info()` permitió obtener una visión general de la estructura del conjunto de datos.

Se verificó que el DataFrame contiene **8160 registros** y **8 variables**, con una combinación de datos numéricos (`int64` y `float64`) y categóricos (`object`).

Durante la inspección también se identificó que algunas variables no poseen información completa para todos los registros. En particular, las columnas **monthly_watch_time_mins**, **favorite_genre** y **last_login_date** presentan una cantidad de valores no nulos inferior al total de registros, lo que indica la existencia de valores faltantes.

Asimismo, se observa que la variable **last_login_date** fue interpretada como un dato de tipo `object`, por lo que será necesario analizar posteriormente si corresponde convertirla al tipo de dato fecha (`datetime`).

Estas observaciones constituyen evidencia inicial que servirá para justificar las decisiones de preparación y limpieza del dataset en el siguiente notebook.

## 8. Análisis de valores faltantes

La presencia de valores faltantes es uno de los aspectos más importantes durante la inspección inicial de un conjunto de datos.

En esta sección se identificará la cantidad de valores nulos presentes en cada variable y se calculará el porcentaje que representan respecto del total de registros.

Esta información permitirá evaluar la magnitud del problema y servirá como evidencia para justificar las decisiones de limpieza y preparación de datos en etapas posteriores.

In [ ]:
# Cantidad de valores faltantes por columna
valores_faltantes = df.isnull().sum()

# Porcentaje de valores faltantes
porcentaje_faltantes = (valores_faltantes / len(df)) * 100

# Crear una tabla resumen
faltantes = pd.DataFrame({
    "Valores faltantes": valores_faltantes,
    "Porcentaje (%)": porcentaje_faltantes.round(2)
})

faltantes

,Valores faltantes,Porcentaje (%)
user_id,0,0.00
age,0,0.00
subscription_plan,0,0.00
monthly_watch_time_mins,193,2.37
country,0,0.00
favorite_genre,240,2.94
last_login_date,320,3.92
customer_support_tickets,0,0.00


### Interpretación

El análisis de valores faltantes permitió identificar que únicamente tres variables presentan datos ausentes: **monthly_watch_time_mins**, **favorite_genre** y **last_login_date**.

La variable con mayor cantidad de valores faltantes es **last_login_date**, con **320 registros** (3,92 %), seguida por **favorite_genre**, con **240 registros** (2,94 %), y **monthly_watch_time_mins**, con **193 registros** (2,37 %).

Las demás variables no presentan valores faltantes, por lo que contienen información completa para la totalidad de los registros.

En general, el porcentaje de datos faltantes observado es relativamente bajo. Sin embargo, antes de decidir si corresponde eliminar, imputar o conservar dichos registros, será necesario analizar el contexto de cada variable durante la etapa de preparación y limpieza de datos.

## 9. Identificación de registros duplicados

La presencia de registros duplicados puede afectar la calidad del análisis, ya que un mismo caso podría contabilizarse más de una vez.

En esta sección se verificará si existen filas completamente duplicadas dentro del conjunto de datos. Esta información permitirá determinar si será necesario aplicar algún procedimiento de eliminación durante la etapa de limpieza de datos.

In [ ]:
# Cantidad de registros duplicados
duplicados = df.duplicated().sum()

print(f"Cantidad de registros duplicados: {duplicados}")

Cantidad de registros duplicados: 126


### Interpretación

El análisis permitió identificar la existencia de **126 registros duplicados**, lo que representa aproximadamente el **1,54 %** del total del conjunto de datos.

La presencia de registros duplicados puede generar sesgos en los análisis posteriores, ya que un mismo caso podría contabilizarse más de una vez.

Sin embargo, en esta etapa únicamente se registra la evidencia observada. La decisión de conservar o eliminar dichos registros será evaluada y justificada durante el proceso de limpieza y preparación de datos.

## 10. Estadísticas descriptivas de las variables numéricas

En esta sección se calcularán las principales estadísticas descriptivas de las variables numéricas del conjunto de datos.

Estas medidas permiten conocer el comportamiento general de los datos, identificar posibles valores atípicos y detectar inconsistencias que podrían requerir un tratamiento durante la etapa de preparación de datos.

In [ ]:
# Estadísticas descriptivas de las variables numéricas
df.describe()

,user_id,age,monthly_watch_time_mins,customer_support_tickets
count,8160.000000,8160.000000,7967.000000,8160.000000
mean,13995.433824,34.096814,1107.346153,1.800980
std,2310.810660,14.511304,5310.442604,11.334969
min,10000.000000,-5.000000,-120.000000,-1.000000
25%,11987.750000,25.000000,489.200000,0.000000
50%,13998.500000,33.000000,757.400000,1.000000
75%,15997.250000,42.000000,1045.700000,1.000000
max,17999.000000,150.000000,99999.000000,150.000000


### Interpretación

Las estadísticas descriptivas permitieron obtener una visión general del comportamiento de las variables numéricas del conjunto de datos.

En términos generales, las variables presentan valores coherentes con el contexto del problema. Sin embargo, durante la inspección se detectaron algunas inconsistencias que deberán analizarse en la etapa de preparación de datos.

En la variable **age** se observaron edades mínimas de **-5 años** y máximas de **150 años**, valores que resultan poco plausibles. Asimismo, la variable **monthly_watch_time_mins** presenta tiempos de visualización negativos y un valor máximo de **99999 minutos**, mientras que **customer_support_tickets** registra un valor mínimo de **-1**.

Estas observaciones constituyen evidencia de posibles errores de carga o registros atípicos. No obstante, en esta etapa únicamente se documentan las inconsistencias detectadas; las decisiones de tratamiento serán justificadas posteriormente durante el proceso de limpieza y preparación del conjunto de datos.

## 11. Análisis de las variables categóricas

En esta sección se analizarán las variables categóricas del conjunto de datos.

El objetivo es identificar las diferentes categorías presentes en cada variable y detectar posibles inconsistencias en la escritura, diferencias de formato o valores atípicos que puedan afectar el análisis posterior.

Esta inspección permitirá justificar futuras decisiones de normalización durante la etapa de preparación de datos.

In [ ]:
# Variables categóricas
variables_categoricas = [
    "subscription_plan",
    "country",
    "favorite_genre"
]

# Mostrar los valores únicos de cada variable
for variable in variables_categoricas:

    print("="*60)
    print(f"Variable: {variable}")
    print("="*60)

    print(df[variable].value_counts(dropna=False))

    print("\n")

Variable: subscription_plan
subscription_plan
Básico       3450
Estándar     2711
Premium      1519
basico         60
BASICO         52
Basic          52
básico         50
Std            48
Estándar       46
estandar       36
STANDARD       34
Premium        31
PREMIUM        26
Premiun        23
premium        22
Name: count, dtype: int64


Variable: country
country
Brasil        1132
Chile         1132
México        1129
Uruguay       1124
Perú          1120
Colombia      1116
Argentina     1087
colombia        27
méxico          25
uruguay         24
Brazil          21
COL             19
CHL             18
URY             17
argentina       16
MEX             16
Chile           16
PER             16
Peru            15
BRA             15
chile           15
Mexico          15
brasil          13
perú            12
ARG             10
Argentina       10
Name: count, dtype: int64


Variable: favorite_genre
favorite_genre
Comedia        1112
Drama          1105
Documental     1095
Romance 

### Interpretación

El análisis de las variables categóricas permitió identificar diversas inconsistencias en la forma en que fueron registrados los datos.

En la variable **subscription_plan** se observaron diferencias de escritura asociadas al uso de mayúsculas y minúsculas, ausencia de acentos, abreviaturas, traducciones al inglés y errores tipográficos.

La variable **country** presenta distintas formas de representar un mismo país, incluyendo nombres en español, inglés, abreviaturas y diferencias de capitalización.

Por su parte, la variable **favorite_genre** evidencia inconsistencias similares, además de traducciones entre español e inglés y diferencias ortográficas entre categorías equivalentes.

Estas observaciones indican que será necesario realizar un proceso de normalización durante la etapa de preparación de datos con el objetivo de unificar categorías equivalentes y garantizar la consistencia del análisis.

## 12. Resumen de problemas detectados

A partir de la inspección inicial realizada en las secciones anteriores, se identificaron diferentes aspectos relacionados con la calidad de los datos.

La siguiente tabla resume los principales problemas detectados, la evidencia observada y una posible línea de trabajo para la etapa de preparación y limpieza del conjunto de datos.

Es importante destacar que las acciones propuestas no constituyen decisiones definitivas, sino posibles alternativas que deberán analizarse y justificarse durante el Notebook 02.

| Variable | Problema detectado | Evidencia | Posible tratamiento |
|----------|--------------------|-----------|---------------------|
| monthly_watch_time_mins | Valores faltantes | 193 registros (2,37 %) | Analizar imputación o eliminación de registros |
| favorite_genre | Valores faltantes | 240 registros (2,94 %) | Evaluar estrategia de tratamiento según el contexto |
| last_login_date | Valores faltantes | 320 registros (3,92 %) | Analizar imputación o conservación de valores nulos |
| last_login_date | Tipo de dato incorrecto | Interpretada como `object` | Convertir a tipo `datetime` |
| age | Valores fuera de rango | Edad mínima -5 y máxima 150 | Evaluar tratamiento de valores inconsistentes |
| monthly_watch_time_mins | Valores inconsistentes | Tiempo mínimo -120 y máximo 99999 minutos | Analizar valores atípicos |
| customer_support_tickets | Valores inconsistentes | Valor mínimo -1 | Verificar consistencia de los registros |
| subscription_plan | Categorías inconsistentes | Diferencias de escritura, idioma y abreviaturas | Normalizar categorías |
| country | Categorías inconsistentes | Países escritos de distintas formas | Unificar nombres |
| favorite_genre | Categorías inconsistentes | Diferencias de idioma, mayúsculas y ortografía | Normalizar categorías |
| Dataset | Registros duplicados | 126 registros duplicados (1,54 %) | Verificar si corresponde eliminarlos |

## 13. Conclusiones de la inspección inicial

La inspección inicial permitió comprender la estructura general del conjunto de datos e identificar los principales aspectos relacionados con su calidad.

Durante el análisis se detectaron valores faltantes, registros duplicados, inconsistencias en los tipos de datos, valores numéricos fuera de rango y diferencias en la representación de categorías equivalentes.

Estas observaciones constituyen evidencia suficiente para justificar la necesidad de realizar una etapa de preparación y limpieza de datos antes de avanzar con el análisis exploratorio.

En consecuencia, el siguiente notebook estará orientado a corregir las inconsistencias detectadas mediante un proceso de limpieza documentado y reproducible, registrando cada transformación realizada sobre el conjunto de datos.